File: javascript/Array_Hash/448. Find All Numbers Disappeared in an Array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDisappearedNumbers = function(nums) {
    const numsSet = new Set(nums);
    const range = new Set(Array.from({length: nums.length}, (_, i) => i + 1));
    const res = [...range].filter(num => !numsSet.has(num));
    return res;
};

var findDisappearedNumbers = function(nums) {
    for (let n of nums) {
        i = Math.abs(n) - 1;
        nums[i] = -1 * Math.abs(nums[i]);
    }
    const res = [];
    for (const [i, n] of nums.entries()) {
        if (n > 0) res.push(i + 1);
    }
    return res;
}

File: python/Array_Hash/448. Find All Numbers Disappeared in an Array.py
'''
Time: O(n)
Space: O(1)

Time: O(n)
Space: O(n)
'''

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        # if n is in nums, turn the value at index n into negative
        for n in nums:
            i = abs(n) - 1  # abs: if n has already been turned into negative
            nums[i] = -1 * abs(nums[i])

        # check
        res = []
        for i, n in enumerate(nums):
            if n < 0:
                res.append(i + 1)
        return res
    
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        return list(set(range(1,len(nums)+1)) - set(nums))
